### PR TITLE
adding transaction management to DAO (GDEV-1151)

### PR DESCRIPTION
**Pull request by KerstenBreuer**
For squash and merge:

```
The DAOs returned by the DaoFactoryProtocol now have
a `with_transaction` method that returns an async context
manager for managing transactions.

Thereby, upon __aenter__ it returns a transactionally scoped DAO.
This recursive structure supports having nested transactions.
Moreover, a provider might implement the DAO operations
performed with and without transactions in a different way
(which is for instance useful for MongoDB, which operates
much faster without than with transactions).

Also updated the find API by splitting it into two separate 
methods find_all and find_one. The first for getting all hits,
the second one for getting a single hit.
Option to return the newest and oldest hit where removed
(there is no easy and generic way to achieve that in 
MongoDB).

Bump version to 0.3.0.
```

### CODE REVIEW

File: hexkit/protocols/dao.py [hunk 1]
```diff
@@ -124,56 +122,47 @@ async def delete(self, *, id_: str) -> None:
...
+            Returns a hit in the form of the respective DTO model or None if no hit
+            was found.
 
         Raises:
-            NoHitsFoundError:
-                Raised when no hits where found when used in "newest", "oldest", or
-                "single" mode. When using the "all" mode, zero hits will not cause an
-                exception but simply result in an empty list beeing returned.
             MultpleHitsFoundError:
                 Raised when obtaining more than one hit when using the "single" mode.
```
Comment (Cito) [inline hexkit/protocols/dao.py:148]:
```suggestion
                Raised when obtaining more than one hit.
```

File: hexkit/protocols/dao.py [hunk 2]
```diff
@@ -184,6 +173,21 @@ class DaoSurrogateId(DaoCommons[Dto], typing.Protocol[Dto, DtoCreation_contra]):
      creation of new resources, is needed.
```
Comment (Cito) [inline hexkit/protocols/dao.py:173]: "is needed" -> i**f** needed?

**Review by Cito — APPROVED** [review]
LGTM. We can still revise the API when implementing the Mongo provider.

File: hexkit/protocols/dao.py (hunk 2 above)
Comment (KerstenBreuer) [reply hexkit/protocols/dao.py:168]: Don't know what was written there :D deleted it

**Pull request merged**